### PR TITLE
Fix scroll restoration when redirecting in an action

### PR DESCRIPTION
**Pull request by jrakotoharisoa**
This Pull request fix issue #9577

### CODE REVIEW

File: packages/router/router.ts [hunk 1]
```diff
@@ -771,6 +771,10 @@ export function createRouter(init: RouterInit): Router {
         )
       : state.loaderData;
 
+    // Don't restore on submission navigations when no redirection
+    let preventScrollRestoration =
+      state.navigation.formData &&
+      location.pathname + location.search === state.navigation.formAction;
```
[the remark this answers is not in the document]
Comment (jrakotoharisoa) [reply packages/router/router.ts]: yes, I've made the change

File: packages/router/__tests__/router-test.ts [hunk 2]
```diff
@@ -6485,6 +6485,37 @@ describe("a router", () => {
...
+      let nav1 = await t.navigate("/tasks", {
+        formMethod: "post",
+        formData: createFormData({}),
+      });
+      const nav2 = await nav1.actions.tasks.redirectReturn("/");
+      await nav2.loaders.index.resolve("INDEX");
+      expect(t.router.state.restoreScrollPosition).not.toBe(false);
+      expect(t.router.state.preventScrollReset).toBe(false);
+    });
+
```
Comment (brophdawg11) [inline packages/router/__tests__/router-test.ts:6518]:
```suggestion
    it("restores scroll on submissions that redirect to the same location", async () => {
      let t = setup({
        routes: SCROLL_ROUTES,
        initialEntries: ["/tasks"],
        hydrationData: {
          loaderData: {
            index: "INDEX_DATA",
          },
        },
      });

      expect(t.router.state.restoreScrollPosition).toBe(false);
      expect(t.router.state.preventScrollReset).toBe(false);
      // We were previously on tasks at 100
      let positions = { "/tasks": 100 };
      // But we've scrolled up to 50 to submit.  We'll save this overtop of
      // the 100 when we start this submission navigation and then restore to
      // 50 below
      let activeScrollPosition = 50;
      t.router.enableScrollRestoration(
        positions,
        () => activeScrollPosition,
        (l) => l.pathname
      );

      let nav1 = await t.navigate("/tasks", {
        formMethod: "post",
        formData: createFormData({}),
      });
      const nav2 = await nav1.actions.tasks.redirectReturn("/tasks");
      await nav2.loaders.tasks.resolve("TASKS");
      expect(t.router.state.restoreScrollPosition).toBe(50);
      expect(t.router.state.preventScrollReset).toBe(false);
    });

    it("restores scroll on submissions that redirect to new locations", async () => {
      let t = setup({
        routes: SCROLL_ROUTES,
        initialEntries: ["/tasks"],
        hydrationData: {
          loaderData: {
            index: "INDEX_DATA",
          },
        },
      });

      expect(t.router.state.restoreScrollPosition).toBe(false);
      expect(t.router.state.preventScrollReset).toBe(false);
      let positions = { "/": 50, "/tasks": 100 };
      let activeScrollPosition = 0;
      t.router.enableScrollRestoration(
        positions,
        () => activeScrollPosition,
        (l) => l.pathname
      );

      let nav1 = await t.navigate("/tasks", {
        formMethod: "post",
        formData: createFormData({}),
      });
      const nav2 = await nav1.actions.tasks.redirectReturn("/");
      await nav2.loaders.index.resolve("INDEX");
      expect(t.router.state.restoreScrollPosition).toBe(50);
      expect(t.router.state.preventScrollReset).toBe(false);
    });
    
```

File: packages/router/router.ts [hunk 3]
```diff
@@ -771,6 +771,9 @@ export function createRouter(init: RouterInit): Router {
         )
       : state.loaderData;
 
+    // Don't restore on submission navigations when no redirection
+    let preventScrollRestoration =
+      state.navigation.formData && location.state?._isRedirect !== true;
```
Comment (brophdawg11) [inline packages/router/router.ts:776]: // Don't restore on submission navigations unless they redirect
    let restoreScrollPosition =
      state.navigation.formData && location.state?._isRedirect !== true
        ? false
        : getSavedScrollPosition(location, newState.matches || state.matches);

File: packages/router/router.ts [hunk 4]
```diff
@@ -780,8 +783,7 @@ export function createRouter(init: RouterInit): Router {
       initialized: true,
       navigation: IDLE_NAVIGATION,
       revalidation: "idle",
-      // Don't restore on submission navigations
-      restoreScrollPosition: state.navigation.formData
+      restoreScrollPosition: preventScrollRestoration
         ? false
         : getSavedScrollPosition(location, newState.matches || state.matches),
```
Comment (brophdawg11) [inline packages/router/router.ts:788]:
```suggestion
      restoreScrollPosition,
```

File: packages/router/router.ts (hunk 3 above)
Comment (brophdawg11) [inline packages/router/router.ts:776]:
```suggestion
    // Don't restore on submission navigations unless they redirect
    let restoreScrollPosition =
      state.navigation.formData && location.state?._isRedirect !== true
        ? false
        : getSavedScrollPosition(location, newState.matches || state.matches);

```

**Pull request merged**